src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs

```diff
@@ -239,5 +339,223 @@ private CspParameters GetPrivateKeyCsp()
                 }
             }
         }
+
+        private unsafe ICertificatePal CopyWithPersistedCngKey(CngKey cngKey)
+        {
+            if (string.IsNullOrEmpty(cngKey.KeyName))
+            {
+                return null;
+            }
+
+            // Make a new pal from bytes.
+            CertificatePal pal = (CertificatePal)FromBlob(RawData, SafePasswordHandle.InvalidHandle, X509KeyStorageFlags.PersistKeySet);
+
+            CngProvider provider = cngKey.Provider;
+            string keyName = cngKey.KeyName;
+            bool machineKey = cngKey.IsMachineKey;
+
+            // CAPI RSA_SIGN keys won't open correctly under CNG without the key number being specified, so
+            // check to see if we can figure out what key number it needs to re-open.
+            int keySpec = GuessKeySpec(provider, keyName, machineKey, cngKey.AlgorithmGroup);
+
+            CRYPT_KEY_PROV_INFO keyProvInfo = new CRYPT_KEY_PROV_INFO();
+
+            fixed (char* keyNamePtr = cngKey.KeyName)
+            fixed (char* provNamePtr = cngKey.Provider.Provider)
+            {
+                keyProvInfo.pwszContainerName = keyNamePtr;
+                keyProvInfo.pwszProvName = provNamePtr;
+                keyProvInfo.dwFlags = machineKey ? CryptAcquireContextFlags.CRYPT_MACHINE_KEYSET : 0;
+                keyProvInfo.dwKeySpec = keySpec;
+
+                if (!Interop.crypt32.CertSetCertificateContextProperty(
+                    pal._certContext,
+                    CertContextPropId.CERT_KEY_PROV_INFO_PROP_ID,
+                    CertSetPropertyFlags.None,
+                    &keyProvInfo))
+                {
+                    pal.Dispose();
+                    throw Marshal.GetLastWin32Error().ToCryptographicException();
+                }
+            }
+
+            return pal;
+        }
+
+        private static int GuessKeySpec(
+            CngProvider provider,
+            string keyName,
+            bool machineKey,
+            CngAlgorithmGroup algorithmGroup)
+        {
+            if (provider == CngProvider.MicrosoftSoftwareKeyStorageProvider ||
+                provider == CngProvider.MicrosoftSmartCardKeyStorageProvider)
+            {
+                // Well-known CNG providers, keySpec is 0.
+                return 0;
+            }
+
+            const int NTE_BAD_KEYSET = unchecked((int)0x80090016);
+
+            try
+            {
+                CngKeyOpenOptions options = machineKey ? CngKeyOpenOptions.MachineKey : CngKeyOpenOptions.None;
```

Comment:
Export is only ever called on the key that was provided; so it'll use whatever silent-or-not policy that key was opened with.  But Export is only called when the type isn't known to us; so someone would have had to have opened a persisted+protected key using a class that isn't one of ours... so we wouldn't know how to open it in silent mode anyways :smile:.